### DISCUSSION TITLE
StreamingResponse leads to empty request content for client

### DISCUSSION THREAD

**Discussion by jackhwolf** [fastapi/fastapi#7117] [Questions] [answered]
I have an api in which the server sends a stream of data to the client. On the server-side, the stream object is populated. I can verify this by writing to a file on the server before sending. But when the client receives the stream and tries to write it to memory, it always results in an empty file. Here is my code

TL;DR: Stream is not empty when sent from api, but is empty when client recieves and tries to write

### Server-side - sending `io.BytesIO` object
```
from io import BytesIO
from fastapi import FastAPI
app = FastAPI()

# capture image and store in io.BytesIO stream
def pic_to_stream(self):
        io = BytesIO()
        self.c.capture(io, 'jpeg')  # populate stream with image
        return io

# api endpoint to send stream
@app.get('/dropcam/stream')
def dc_stream():
    stream, ext = cam.pic_to_stream()
    # with open('example.ext', 'wb') as f:
    #     f.write(stream.getbuffer())  # this results in non-empty file on server
    return StreamingResponse(stream, media_type=ext)
```

### Client-side - receiving `io.BytesIO` object and trying to write
```
import requests

def recieve_stream():
    url = 'http://abc.x.y.z/dropcam/stream'
    local_filename = 'client/foo_stream.jpeg'
    with requests.get(url, stream=True) as r:
        with open(local_filename, 'wb') as f:
            f.write(r.content) # this results in an empty file on client
```

I'm sorry if this is basic. I've been working on this for a while with no luck. Does anyone have any ideas? I have looked at [these][1] [links][2] and [more][3]. Thank you!

  [1]: https://github.com/psf/requests/issues/4227
  [2]: https://stackoverflow.com/questions/45236801/downloading-file-using-requests-module-creates-an-empty-file
  [3]: https://stackoverflow.com/questions/16694907/download-large-file-in-python-with-requests

**Comment by Mause** [comment]
You need to rewind the file using .tell(0)

Though why not just return the file using FileResponse?

**Comment by jackhwolf** [comment, marked as answer]
Thank you so much! I ended up using .seek(0) on the io object before returning it in StreamingResponse.

The reason I'm using StreamingResponse is that I am working on a raspberry pi and wanted to save disk space and r/w time. I also wanted to play around with it just for fun.

**Comment by Mause** [comment]
Ah yep sorry, wrong method 😅

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! :clap: :bow: 

Thanks for reporting back and closing the issue @jackhwolf  :+1:
